storage/innobase/handler/ha_innodb.cc

```diff
@@ -17080,6 +17080,52 @@ int innobase_rollback_by_xid(handlerton* hton, XID* xid)
 	}
 }
 
+/**
+This function is used to rollback one X/Open XA distributed transaction
+which is in the prepared state asynchronously.
+
+It only set the transaction's status to ACTIVE and persist the status.
+The transaction will be rolled back by background rollback thread.
+
+@param[in] hton InnoDB handlerton
+@param[in] xid X/Open XA transaction identification
+
+@return 0 or error number */
+int innobase_recover_rollback_by_xid(handlerton *hton, XID *xid)
+{
+	DBUG_ASSERT(hton == innodb_hton_ptr);
+	DBUG_EXECUTE_IF("innobase_xa_fail", return XAER_RMFAIL;);
+	DBUG_EXECUTE_IF("crash_before_async_binlog_recovery", DBUG_SUICIDE(););
+
+	if (high_level_read_only) {
+		return(XAER_RMFAIL);
+	}
+
+	trx_t *trx = trx_get_trx_by_xid(xid);
+	if (trx == NULL) {
+		return(XAER_RMFAIL);
+	}
+
+  // ddl should not be rolled back through recovery
+  ut_ad(!trx->dict_operation);
+
+#ifdef WITH_WSREP
+	/* If a wsrep transaction is being rolled back during
+	the recovery, we must clear the xid in order to avoid
+	writing serialisation history for rolled back transaction. */
+	if (wsrep_is_wsrep_xid(&trx->xid)) {
+		trx->xid.null();
+	}
+#endif /* WITH_WSREP */
```

Comment:
For other transactions, this is done as part of `trx_get_trx_by_xid_callback()`, protected by `trx->mutex`. There seems to be a potential race condition with this assignment: Multiple concurrent threads could get the same object returned by `trx_get_trx_by_xid()` in case `wsrep_is_wsrep_xid(&trx->xid)` holds. Can this condition ever hold in this code path? Could we simply write the following?
```c++
#ifdef WITH_WSREP
  ut_ad(!wsrep_is_wsrep_xid(&trx->xid));
#endif
```

Reply:
done